src/parser/Parser.ts

```diff
@@ -184,1368 +188,1390 @@ export class Parser {
+                    throw this.addError(
+                        {
+                            kind: Lexeme.Identifier,
+                            text: arg.name.text,
+                            isReserved: ReservedWords.has(arg.name.text),
+                            location: arg.location,
+                        },
+                        `Argument '${arg.name.text}' has no default value, but comes after arguments with default values`
                     );
                 }
 
-                let func = new Expr.Function(
-                    args,
-                    returnType,
-                    body,
-                    startingKeyword,
-                    endingKeyword
-                );
+                return haveFoundOptional || !!arg.defaultValue;
+            }, false);
 
-                if (isAnonymous) {
-                    return func;
-                } else {
-                    // only consume trailing newlines in the statement context; expressions
-                    // expect to handle their own trailing whitespace
-                    while (match(Lexeme.Newline));
-                    return new Stmt.Function(name!, func);
-                }
-            } finally {
-                functionDeclarationLevel--;
+            this.consume(
+                `Expected newline or ':' after ${functionType.text} signature`,
+                Lexeme.Newline,
+                Lexeme.Colon
+            );
+            //support ending the function with `end sub` OR `end function`
+            let body = this.block(Lexeme.EndSub, Lexeme.EndFunction);
+            if (!body) {
+                throw this.addError(
+                    this.peek(),
+                    `Expected 'end ${functionType.text}' to terminate ${functionType.text} block`
+                );
             }
-        }
-
-        function signatureArgument(): Argument {
-            if (!check(Lexeme.Identifier)) {
-                throw addError(
-                    peek(),
-                    `Expected argument name, but received '${peek().text || ""}'`
+            // consume 'end sub' or 'end function'
+            let endingKeyword = this.advance();
+            let expectedEndKind = isSub ? Lexeme.EndSub : Lexeme.EndFunction;
+
+            //if `function` is ended with `end sub`, or `sub` is ended with `end function`, then
+            //add an error but don't hard-fail so the AST can continue more gracefully
+            if (endingKeyword.kind !== expectedEndKind) {
+                this.addError(
+                    endingKeyword,
+                    `Expected 'end ${functionType.text}' to terminate ${functionType.text} block`
                 );
             }
 
-            let name = advance();
-            let type: ValueKind = ValueKind.Dynamic;
-            let typeToken: Token | undefined;
-            let defaultValue;
+            let func = new Expr.Function(args, returnType, body, startingKeyword, endingKeyword);
```

Comment:
I think prettier reformatted this line because it now fits on one line since it's de-indented